presto-main/src/test/java/com/facebook/presto/sql/planner/BenchmarkPlanner.java

```diff
@@ -0,0 +1,161 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.testing.LocalQueryRunner;
+import com.facebook.presto.tpch.TpchConnectorFactory;
+import com.google.common.base.Charsets;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.io.Resources;
+import io.airlift.tpch.Customer;
+import org.intellij.lang.annotations.Language;
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.BenchmarkMode;
+import org.openjdk.jmh.annotations.Fork;
+import org.openjdk.jmh.annotations.Measurement;
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.annotations.OutputTimeUnit;
+import org.openjdk.jmh.annotations.Param;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.TearDown;
+import org.openjdk.jmh.annotations.Warmup;
+import org.openjdk.jmh.runner.Runner;
+import org.openjdk.jmh.runner.options.Options;
+import org.openjdk.jmh.runner.options.OptionsBuilder;
+import org.openjdk.jmh.runner.options.VerboseMode;
+import org.openjdk.jmh.runner.options.WarmupMode;
+
+import java.io.IOException;
+import java.net.URL;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+import static com.facebook.presto.testing.TestingSession.testSessionBuilder;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+
+@SuppressWarnings("MethodMayBeStatic")
+@State(Scope.Benchmark)
+@OutputTimeUnit(TimeUnit.MILLISECONDS)
+@Warmup(iterations = 5)
+@Fork(1)
+@Measurement(iterations = 20)
+@BenchmarkMode(Mode.AverageTime)
+public class BenchmarkPlanner
+{
+    @SuppressWarnings("FieldMayBeFinal")
+    @State(Scope.Benchmark)
+    public static class BenchmarkData
+    {
+        @Param({"", "EXPLAIN"})
+        private String queryPrefix = "";
+
+        @Param({"true", "false"})
+        private String iterativeOptimizerEnabled = "true";
+
+        private LocalQueryRunner queryRunner;
+        private List<String> queries;
+        private Session session;
+
+        @Setup
+        public void setup()
+        {
+            String tpch = "tpch";
+
+            session = testSessionBuilder()
+                    .setCatalog(tpch)
+                    .setSchema("sf1000")
+                    .setSystemProperty("iterative_optimizer_enabled", iterativeOptimizerEnabled)
+                    .build();
+
+            queryRunner = new LocalQueryRunner(session);
+            queryRunner.createCatalog(tpch, new TpchConnectorFactory(4), ImmutableMap.of("tpch.column-naming", "standard"));
+
+            queries = IntStream.rangeClosed(1, 22)
```

Comment:
Benchmarking planning for all queries is probably too big unit of measure. After all, this is rather micro benchmark framework. With such big test unit it will be hard to catch regressions for individual queries. Instead maybe it would be better to have query number as a `@Param`, so we can see planning time for each query.

Reply:
I like the idea. However, it's harder to eye-compare 21 results than just 1 aggregated result. We can also do both at once: have `@Param` selecting query number and special value for all-of-them

Reply:
My point was to provide a general notion of planner performance. That way I will be able to measure general impact of things we introduce into a planner. Notice the commit message where I measured how big part of planning takes optimizer (it is currently about 60%).

Reply:
> Notice the commit message where I measured how big part of planning takes optimizer (it is currently about 60%).

How did you came to that conclusion?

Reply:
Normally tpch queries were executed in about 390ms. When I changed the planner code to run optimizer rules two times it grew to about 645ms. Which is 255ms more. 255/390 gives about 65%. I know it is not exact nor scientific method, but it gives a notion how things work.

Reply:
> My point was to provide a general notion of planner performance.
I guess we could live with that for now.

However:
There is `createPlan` method that accepts Stage: `LocalQueryRunner#createPlan(com.facebook.presto.Session, java.lang.String, java.util.List<com.facebook.presto.sql.planner.optimizations.PlanOptimizer>, com.facebook.presto.sql.planner.LogicalPlanner.Stage)`. Stages are `CREATED, OPTIMIZED, OPTIMIZED_AND_VALIDATED`. Maybe this should be benchmark `@Param`?